numpy/core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -1724,6 +1741,10 @@ mapiter_@name@(PyArrayMapIterObject *mit)
                         NPY_AUXDATA_FREE(transferdata);
                         return -1;
                     }
+                    if(subspace_ptrs[0] == self_ptr &&
+                       subspace_ptrs[1] == mit->extra_op_ptrs[0]) {
+                        skip <<= 1;
```

Comment:
For the second part of your question: shifting stops after skip == 4 -- if skip started with 1. Otherwise skip started with 0 and shifting never do anything. So it won't overflow.  

The first part of your question I don't really understand.